go/mcap/cmd/doctor.go

```diff
@@ -0,0 +1,178 @@
+package cmd
+
+import (
+	"os"
+	"io"
+	"bytes"
+	"fmt"
+	"encoding/binary"
+
+	"github.com/fatih/color"
+	"github.com/foxglove/mcap/go/libmcap"
+	"github.com/spf13/cobra"
+)
+
+type mcapRecordReader struct {
+	reader       io.Reader
+}
+
+type RawRecord struct {
+	OpCode libmcap.OpCode
+	Bytes []byte
+}
+
+func (this *mcapRecordReader) Warn(format string, v ...interface{}) {
+	color.Yellow(format, v...)
+}
+
+func (this *mcapRecordReader) Error(format string, v ...interface{}) {
+	color.Red(format, v...)
+}
+
+func (this *mcapRecordReader) Fatal(v ...interface{}) {
+	color.Set(color.FgRed)
+	fmt.Println(v...)
+	color.Unset()
+	os.Exit(1)
+}
+
+func (this *mcapRecordReader) Fatalf(format string, v ...interface{}) {
+	color.Red(format, v...)
+	os.Exit(1)
+}
+
+
+func (this *mcapRecordReader) ConsumeMagic() {
+	magicBytes := make([]byte, len(libmcap.Magic))
+
+	bytesRead, err := this.reader.Read(magicBytes)
+	if err != nil {
+		this.Fatal(err)
+	}
+	if (bytesRead != len(libmcap.Magic)) {
+		this.Fatal("Initial MAGIC bytes missing")
+	}
+	if !bytes.Equal(magicBytes, libmcap.Magic) {
+		this.Fatal("Invalid MAGIC")
+	}
+}
+
+func (this *mcapRecordReader) NextRecord() (*RawRecord) {
+	opcodeAndLength := make([]byte, 9)
+
+	bytesRead, err := this.reader.Read(opcodeAndLength)
+	if err != nil {
+		this.Fatal(err)
+	}
+	if (bytesRead != 9) {
+		this.Fatal("No more bytes to read. Expecting Record Opcode byte and Record Length.")
+	}
+
+	opcode := libmcap.OpCode(opcodeAndLength[0])
+	recordLen := binary.LittleEndian.Uint64(opcodeAndLength[1:9])
+
+	recordBytes := make([]byte, recordLen)
+	bytesRead, err = this.reader.Read(recordBytes)
+	if err != nil {
+		this.Fatal(err)
+	}
+	if (uint64(bytesRead) != recordLen) {
+		this.Fatalf("Could not read record payload. Wanted %d bytes, got %d.", )
+	}
+
+	return &RawRecord {
+		OpCode: opcode,
+		Bytes: recordBytes,
+	}
+}
+
+func NewMcapRecordReader(reader io.Reader) *mcapRecordReader {
+	return &mcapRecordReader{
+		reader:      reader,
+	}
+}
+
+
+func main(cmd *cobra.Command, args []string) {
+	if len(args) != 1 {
+		fmt.Println("An mcap file argument is required.")
+		os.Exit(1)
+	}
+
+	fmt.Printf("Examining %s\n", args[0])
+	file, err := os.Open(args[0])
+	if err != nil {
+		fmt.Print(err)
+		os.Exit(1)
+	}
+
+	recordReader := NewMcapRecordReader(file)
+
+	recordReader.ConsumeMagic()
+
+	var previousOpcode *libmcap.OpCode = nil
+	
+	for {
+		var rawRecord = recordReader.NextRecord()
+
+		if (previousOpcode == nil && rawRecord.OpCode != libmcap.OpHeader) {
+			recordReader.Error("First record MUST be a header")
+		}
+		
+		switch rawRecord.OpCode {
+		case libmcap.OpHeader:
+			// warn if header doesn't contain a library field - this is encouraged
+			// error if header field has a profile that is not well known and is not prefixed by "x-"
+			break
```

Comment:
break in go is not required to break out of a switch. This will actually break out of the for loop. The breaks should just be removed; the code will exit after it finds a case, unless explicitly told to `fallthrough`.